compiler/src/dotty/tools/dotc/typer/Checking.scala

```diff
@@ -75,8 +75,12 @@ object Checking {
     }.getOrElse(TypeTree(tparam.paramRef))
     val orderedArgs = if (hasNamedArg(args)) tparams.map(argNamed) else args
     val bounds = tparams.map(_.paramInfoAsSeenFrom(tree.tpe).bounds)
-    def instantiate(bound: Type, args: List[Type]) =
-      HKTypeLambda.fromParams(tparams, bound).appliedTo(args)
+    def instantiate(bound: Type, args: List[Type]) = tparams match {
+      case (_: Symbol) :: _ if args.exists(_.isInstanceOf[TypeBounds]) =>
+        bound.substApprox(tparams.asInstanceOf[List[Symbol]], args)
```

Comment:
SubstApproxMap will start with `variance = 1` but when we call `instantiate(bounds.lo, argTypes)` surely we should start with `variance = -1` ? Maybe we should change `instantiate` to take a `TypeBounds` as input and as output.

Reply:
Also maybe leave a comment to note that we're not certain of the correctness here ?